lib/kinesis_client/stream/shard/producer.ex

```diff
@@ -131,49 +135,29 @@ defmodule KinesisClient.Stream.Shard.Producer do
   end
 
   @impl GenStage
-  def handle_info({:ack, _ref, [], failed_msgs}, state) do
-    Logger.debug("Retrying #{length(failed_msgs)} failed messages")
-
-    state =
-      case state.shard_closed_timer do
-        nil ->
-          state
-
-        timer ->
-          Process.cancel_timer(timer)
-          %{state | shard_closed_timer: nil}
-      end
-
-    {:noreply, failed_msgs, state}
+  def handle_info(msg, state) do
+    Logger.debug("ShardConsumer.Producer got an unhandled message #{inspect(msg)}")
+    {:noreply, [], state}
   end
 
-  @impl GenStage
-  def handle_info({:ack, _ref, successful_msgs, failed_msgs}, state) do
-    %{metadata: %{"SequenceNumber" => checkpoint}} = successful_msgs |> Enum.reverse() |> hd()
-
-    :ok = update_checkpoint(state, checkpoint)
+  def get_top_checkpoint(successful_msgs, failed_msgs) do
+    success_checkpoint = get_top_checkpoint(successful_msgs)
+    failed_checkpoint = get_top_checkpoint(failed_msgs)
+    checkpoint = Enum.sort([success_checkpoint, failed_checkpoint], :desc) |> hd()
```

Comment:
Nit I believe this is equivalent and I think a little clearer:

```suggestion
    checkpoint = Enum.max([success_checkpoint, failed_checkpoint])
```